src/com/google/javascript/jscomp/ProcessCommonJSModules.java

```diff
@@ -338,371 +287,790 @@ private void visitRequireCall(NodeTraversal t, Node require, Node parent) {
-      exportRefs.clear();
+      String moduleName = modulePath.toModuleName();
 
       // Add goog.provide calls.
       if (reportDependencies) {
-        CompilerInput ci = t.getInput();
         ci.addProvide(moduleName);
       }
-      script.addChildToFront(IR.exprResult(
+      this.script.addChildToFront(IR.exprResult(
           IR.call(IR.getprop(IR.name("goog"), IR.string("provide")),
-              IR.string(moduleName))).useSourceInfoIfMissingFromForTree(script));
+              IR.string(moduleName))).useSourceInfoIfMissingFromForTree(this.script));
 
-      compiler.reportCodeChange();
+      // The default declaration for the goog.provide is a constant so
+      // we need to declare the variable if we have more than one
+      // assignment to module.exports or those assignments are not
+      // at the top level.
+      //
+      // If we assign to the variable more than once or all the assignments
+      // are properties, initialize the variable as well.
+      int directAssignmentsAtTopLevel = 0;
+      int directAssignments = 0;
+      for (Node export : moduleExports) {
+        if (NodeUtil.getEnclosingScript(export) == null) {
+          continue;
+        }
+
+        Node base = getBaseQualifiedNameNode(export);
+        if (base == export && export.getParent().isAssign()) {
+          Node rValue = NodeUtil.getRValueOfLValue(export);
+          if (rValue == null || !rValue.isObjectLit()) {
+            directAssignments++;
+            if (export.getParent().getParent().isExprResult() &&
+                NodeUtil.isTopLevel(export.getParent().getParent().getParent())) {
+              directAssignmentsAtTopLevel++;
+            }
+          }
+        }
+      }
+
+      if (directAssignmentsAtTopLevel > 1 || (directAssignmentsAtTopLevel == 0 && directAssignments > 0) ||
+          directAssignments == 0) {
+        int totalExportStatements = this.moduleExports.size() + this.exports.size();
+        Node initModule = IR.var(IR.name(moduleName));
+        if (directAssignments < totalExportStatements) {
+          initModule.getFirstChild().addChildToFront(IR.objectlit());
+
+          // If all the assignments are property exports, initialize the
+          // module as a namespace
+          if (directAssignments == 0) {
+            JSDocInfoBuilder builder = new JSDocInfoBuilder(true);
+            builder.recordConstancy();
+            initModule.setJSDocInfo(builder.build());
+          }
+        }
+        initModule.useSourceInfoIfMissingFromForTree(this.script);
+
+        Node refChild = this.script.getFirstChild();
+        while (refChild.getNext() != null & refChild.getNext().isExprResult() &&
+            refChild.getNext().getFirstChild().isCall() &&
+            ("goog.require".equals(refChild.getNext().getFirstChild().getFirstChild().getQualifiedName()) ||
+             "goog.provide".equals(refChild.getNext().getFirstChild().getFirstChild().getQualifiedName()))) {
```

Comment:
I think these could be simplified with .matchesQualifiedName

If you don't mind the slightly strange name, there's also getFirstFirstChild.